### PR TITLE
Fixed RUBY-156

### PR DESCRIPTION
**Pull request by stamhankar999**
Don't throw out the whole buffer when we see a warning in a frame, since the buffer may contain some/all of the next frame(s).

### CODE REVIEW

File: lib/cassandra/protocol/cql_byte_buffer.rb [hunk 1]
```diff
@@ -78,13 +78,15 @@ def read_long
       def read_double
```
Comment (stamhankar999) [inline lib/cassandra/protocol/cql_byte_buffer.rb]: This whole file has only formatting changes.

**Comment by avalanche123** [comment]
:+1:

File: lib/cassandra/protocol/v3.rb [hunk 2]
```diff
@@ -190,93 +210,117 @@ def actual_decode(buffer, fields, size, code)
 
         def decode_response(opcode, protocol_version, buffer, size, trace_id)
```
Comment (stamhankar999) [inline lib/cassandra/protocol/v3.rb]: The majority of this method has formatting changes. I added a few else blocks to some cases. Noted below.

File: lib/cassandra/protocol/v3.rb [hunk 3]
```diff
@@ -190,93 +210,117 @@ def actual_decode(buffer, fields, size, code)
...
-              original_buffer_length = buffer.length
-              column_specs, paging_state = Coder.read_metadata_v3(buffer)
-
-              if column_specs.nil?
-                consumed_bytes  = original_buffer_length - buffer.length
-                remaining_bytes = CqlByteBuffer.new(buffer.read(size - consumed_bytes - 4))
-                RawRowsResultResponse.new(nil, nil, protocol_version, remaining_bytes, paging_state, trace_id)
-              else
-                RowsResultResponse.new(nil, nil, Coder.read_values_v3(buffer, column_specs), column_specs, paging_state, trace_id)
+            when 0x00 # ERROR
```
Comment (stamhankar999) [inline lib/cassandra/protocol/v3.rb]: The majority of this method has formatting changes. I added a few else blocks to some cases. Noted below.

File: lib/cassandra/protocol/v4.rb [hunk 4]
```diff
@@ -213,119 +232,161 @@ def actual_decode(buffer, fields, size, code)
...
+                        Protocol::Constants::SCHEMA_CHANGE_TARGET_UDT
+                      keyspace = buffer.read_string
+                      name = buffer.read_string
+                    when Protocol::Constants::SCHEMA_CHANGE_TARGET_FUNCTION,
+                        Protocol::Constants::SCHEMA_CHANGE_TARGET_AGGREGATE
+                      keyspace = buffer.read_string
+                      name = buffer.read_string
+                      arguments = buffer.read_string_list
+                    else
+                      raise Errors::DecodingError,
```
Comment (stamhankar999) [inline lib/cassandra/protocol/v4.rb]: This is a new else clause.

File: lib/cassandra/protocol/v4.rb [hunk 5]
```diff
@@ -213,119 +232,161 @@ def actual_decode(buffer, fields, size, code)
           end
         end
 
-        def decode_response(opcode, protocol_version, buffer, size, trace_id, custom_payload, warnings)
+        def decode_response(opcode, protocol_version, buffer, size, trace_id,
+                            custom_payload, warnings)
```
Comment (stamhankar999) [inline lib/cassandra/protocol/v4.rb]: Much like in v3, this method is largely just formatting changes, but I did add some 'else' clauses to some cases where they were missing. Noted below.

File: lib/cassandra/protocol/v4.rb [hunk 6]
```diff
@@ -213,119 +232,161 @@ def actual_decode(buffer, fields, size, code)
...
+                      keyspace = buffer.read_string
+                      name = nil
+                    when Protocol::Constants::SCHEMA_CHANGE_TARGET_TABLE,
+                        Protocol::Constants::SCHEMA_CHANGE_TARGET_UDT,
+                        Protocol::Constants::SCHEMA_CHANGE_TARGET_FUNCTION,
+                        Protocol::Constants::SCHEMA_CHANGE_TARGET_AGGREGATE
+                      keyspace = buffer.read_string
+                      name = buffer.read_string
+                    else
+                      raise Errors::DecodingError,
```
Comment (stamhankar999) [inline lib/cassandra/protocol/v4.rb]: This is a new else clause.

File: spec/cassandra/protocol/v3/decoder_spec.rb [hunk 7]
```diff
@@ -20,7 +20,7 @@
 
 module Cassandra
   module Protocol
-    module V4
+    module V3
```
Comment (stamhankar999) [inline spec/cassandra/protocol/v3/decoder_spec.rb]: Yup, the two decoder tests were actually testing the *other* version.

**Pull request merged**